### ISSUE TITLE
kv: Improve transaction retry error messaging

### ISSUE THREAD

**Issue by awoods187** [cockroachdb/cockroach#41057]
We recently troubleshooted a customer error (https://github.com/cockroachlabs/support/issues/334):
TransactionRetryWithProtoRefreshError:TransactionAbortedError(ABORT_REASON_ABORT_SPAN)
 
This error is not in our documentation and it’s not immediately clear what caused it. In this case, it is likely due to another transaction aborting this transaction which suggests a write contention. The customer expressed frustration with this process: What does this error even mean? Why does it happen? How can I find the information in documentation? What do I do now? 

I went ahead and filed an [issue to improve our docs](https://github.com/cockroachdb/docs/issues/5476) and this issue to improve our error messaging to provide more helpful information to the developer.

**Comment by rmloveland** [comment]
Looking at the descriptive comments in `errors.proto`, it seems that each error message corresponds to some combination of states held by internal mechanisms (timestamp cache, txn record creation, abort spans, etc.).  These mechanisms are of interest to core CockroachDB developers, but not necessarily to users of a SQL API.

In almost every case, the correct response to whichever error type seems to be: "Find the write contention in your application's data access patterns, and remove it if possible.  If you can't remove the contention entirely and are still seeing this error intermittently, add a retry loop to your application".

And this is what our docs basically already say!

Unfortunately, these hyper-specific error messages are in some sense causing users to avoid this pre-existing advice in the docs IMO.  The specificity implies importance - "I must understand this exact error message, it's so specific, there must be a reason they are showing this to me".

Unfortunately even if you do understand that "this one means the timestamp cache does not have a txn ID, but the lease under which the request is evaluated is newer ..." etc., there is nothing you can really do about that from the SQL API.  So you are effectively being misled into thinking some hyper specific action re: this one error code is needed...

But it still boils down to: "you probably have contention, try to remove it.  if you can't entirely, try adding a retry loop".

Re: the actual error messaging, my thesis is there is very little value to the user in providing anything beyond "restart transaction" and a 40001 error code

Open to being corrected on any misunderstandings in the above

**Comment by knz** [comment]
The detail is needed in some internal tests, and also during troubleshooting of unexpected situations. So I think we should not remove the detail entirely.

However, we could make it conditional, and only enable the detail upon explicit request (e.g. a "debug mode" session var).

**Comment by knz** [comment]
@andreimatei  I think historically you've been the most active consumer of these detailed errors. Do you see flaws in the proposal to only expose the detail during tests or troubleshooting?

**Comment by nvanbenschoten** [comment]
I'm very supportive of removing the transaction proto from the error message and slimming it all down. It's way too noisy right now and confuses everyone who looks at it.

However, I do not think we should remove all mention of the retry reason. This information has been critical to our ability to debug retry errors in customer applications countless times and removing it would be a big step back in terms of visibility into customer problems. Adding a debug mode session variable is also not a very good option because it means that we now need to jump through hurdles with the customer just to get back to the point where we currently are in the debugging process. That's a round or two of back and forth before we get anywhere productive.

Even if we just left something like `40001: restart transaction (debug reason: 2)` or just `40001: restart transaction (2)` we'd be a lot better off.

> In almost every case, the correct response to whichever error type seems to be: "Find the write contention in your application's data access patterns, and remove it if possible. If you can't remove the contention entirely and are still seeing this error intermittently, add a retry loop to your application".

This is mostly true, but there's often nuance here that requires us to know the reason for the error in order to help diagnose issues. Relying on this also binds our hands in terms of future changes to the cases where we return retry errors.

**Comment by irfansharif** [comment]
> But it still boils down to: "you probably have contention, try to remove it. if you can't entirely, try adding a retry loop".

Sometimes these retry conditions can arise from internal lease transfers and/or replica movement. There are various other causes of retries in addition to just contention. For these cases, for users who have designed their applications around not having any form of contention whatsoever, our messaging around "retry errors are caused by contention" can be a confusing (something @jasobrown and friends ran into recently).

I'm unsure what's the best course of action here. I agree that generally client code should "simply throw a retry loop at the problem", but I suspect for keener developers understanding the exact cause of the retry would inform them what they'd have to do on their end to minimize the impact (on their p99s or what not). If, say, internal lease transfers or replica movements are a persistent problem, as a developer I would look to pin things down more stringently using zone configs. I'm not sure, but it's at least one instance of where understanding the system-wide cause of a certain retry error would lend to the user trying a few things on their end. What'd ya think?

**Comment by andreimatei** [comment]
I think currently our error message are pretty inscrutable, so they definitely stand to be improved. In particular the text should separate the part that we expect users to read versus the part that only makes sense to a couple of us. The name `TransactionRetryWithProtoRefreshError` makes me chuckle every time.
 But I don't think getting rid of them is the way to go. Different errors do say things that can be important. An uncertain read is different from other types of contention. A lease transfer causing a timestamp cache reset does not even indicate contention but it's still a retriable error (the line for what constitutes a retriable error and what doesn't has historically been quite blurry). A deadlock can be quite actionable for the application. Other reasons for `TransactionAbortedError` are also interesting, and not simply "contention".

Even when it's "just contention", there's still nuances that can be interesting - writes to the same row conflicting, versus more general serializability failures (the former being "write too old").

**Comment by knz** [comment]
I like Nathan's proposal: keep the detail, but only in numeric form. This way we retain the ability to "do something about it" but the casual user will not be tempted to read too much into it.

Andrei what do you think about that?

**Comment by andreimatei** [comment]
My opinion is that people get delighted when they see a message that they understand. I know it makes me feel smart. I think putting a `debug reason: 2` in there will only lead to a search for what that means, and the response to that search, even in the best case, will be generic.
If anything, I'd like *more* information in the errors - for example, if a refresh fails to hide a prior error, we should augment the prior error with the key that caused the refresh to fail. If there's a deadlock, we should put as much info as we can about the locks of the transactions involved. If there's a deferred write too old error, we should remember info about what was deferred. Generally, I think seeing keys/rows in these errors would be a good thing.
Of course, the key is to keep eye on the readability of the messages. I think the current failure is that the string is just gibberish, not that it's too long.

**Comment by andreimatei** [comment]
Btw, the reason listed on slack - `RETRY_ASYNC_WRITE_FAILURE` - is also one that is not about contention. Collecting more specific info when that happens seems hard, but damn would it be useful for someone seeing that error.

**Comment by jasobrown** [comment]
As a user of the database, I admit I agree with most sentiments here. Every time I see `TransactionRetryWithProtoRefreshError`, I think to myself, "I literally no idea WTF that is, nor do my users".

I agree 100% with @andreimatei's statement: "I think the current failure is that the string is just gibberish, not that it's too long." Indeed, length isn't the problem (for advanced users, at least), having useful and intelligible is what is needed. It took me about 12 months before I was really comfortable interpreting those retry error messages, and I had to read the source code before I felt I really grokked it.

@irfansharif:
> for users who have designed their applications around not having any form of contention whatsoever, our messaging around "retry errors are caused by contention" can be a confusing

I understand the reasons why we ran into that error, and how it's difficult to identify it properly. Perhaps that comes down to, as a community, having broader experience with these problems and having additional documentation around it (preferably in the retry docs).

**Comment by rmloveland** [comment]
> My opinion is that people get delighted when they see a message that they understand. I know it makes me feel smart. I think putting a debug reason: 2 in there will only lead to a search for what that means

I agree with this, Andrei - say you were about to deploy your app on CRDB and you get two retry errors with two different codes (`reason: 1` and `reason: 2`.  You would definitely feel compelled to get to the bottom of what each reason means.

In effect this new set of codes would probably not be very different from the current state: paste a magic string into search and see if there is some doc written up somewhere about it.

There is also the issue that Irfan mentioned:

> Sometimes these retry conditions can arise from internal lease transfers and/or replica movement. There are various other causes of retries in addition to just contention.

My dumb question: why do we ever emit retry errors for something other than contention?  this seems like a whole other category of things.  If the system transferred leases or moved replicas, it knows about that, right?  Why does my application need to do the retry?  The whole point of CRDB is that my Java app shouldn't have to know anything about lease transfers and replicas: I talk SQL/JDBC and CockroachDB does the magic to put/get stuff on disk.

I don't expect the database to fix my app's contention-creating logic, but I do expect it to protect me from caring about its internal state as much as possible.

Am I crazy?

**Comment by knz** [comment]
> Am I crazy?

No, just disappointed. As much as we agree with you, the point is that **CockroachDB cannot retry automatically** in certain cases: specifically, all cases where some results have already been communicated to the client.

The core of the issue is that the SQL language is "conversational" - the client can send some stuff to the server during a transaction, receive some results, and then decide to do something else.

There's no way for the server to "simply retry" the stuff done so far, because if there are different results than the first time the client needs to know to perhaps decide to do something else.

**Comment by knz** [comment]
(NB We already do all the retries internally when we know that no result have been transferred to the client yet)

**Comment by irfansharif** [comment]
We now have https://github.com/cockroachdb/docs/issues/5476, thanks to @rmloveland and others.

**Comment by andy-kimball** [comment]
@rmloveland asked that I re-open this issue, after I stumbled into this error message while working on the managed service. Here is some of the conversation.

I've been struggling on the CC side with this really user-unfriendly error message:
```
panic: error committing or rolling back transaction: pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=1c6b273b key=/Table/18125/1/"\x80\x00\x00\a\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"/0 pri=0.04025176 epo=0 ts=1598054224.597738000,2 min=1598054224.597738000,0 seq=17} lock=true stat=PENDING rts=1598054224.597738000,0 wto=false max=1598054224.597738000,0
```
This is not a user-friendly error message, but instead a message intended for the maintainers for the software. As a user, what I want from this error is not a bunch of internal details. What I want is to understand how to get rid of this retry, by understanding which two of my transactions/statements are conflicting with one another. But this unhelpful error message gives me almost no way to do that. Ideally, I want an error message that's like the Golang race condition error, where I can see what two things are conflicting, so I can eliminate one of them.

I bet if we brainstormed, we could figure out a good way to map this into a much better error that would let users understand what's wrong and how to fix it. It's not enough to just say, "just add a transaction retry loop", because that's not what you want to do in many cases. In this case, I have perfect control over my transactions, and can run them differently to reduce contention, if I just knew what was causing that contention.

I realize that "fixing" this error message to let users understand what two transactions are conflicting is probably quite a bit of invasive work, since I don't think we have a good way of tracking user-level transactions down at the KV level. But I think it'd make a huge difference to usability if we could find a good way to do this.

**Comment by knz** [comment]
Regarding the question of better exposing the key that the transaction is anchored in (e.g. "which table?", "which column?")

As long as we embed the string of the txn key in the error message directly, then no, it's too late to make it human-readable. The proper way forward is to make the retry error a structured error; embed the raw bytes as a special error payload; then when the error crosses the sql/kv boundary (coming from kv), rewrite that key using the local SQL codec and table descriptor to make it more human-friendly.
